homeassistant/components/tuya/__init__.py

```diff
@@ -107,86 +122,154 @@ def load_devices(device_list):
                     device_type_list[ha_type] = []
                 device_type_list[ha_type].append(device.object_id())
                 hass.data[DOMAIN]["entities"][device.object_id()] = None
-        for ha_type, dev_ids in device_type_list.items():
-            discovery.load_platform(hass, ha_type, DOMAIN, {"dev_ids": dev_ids}, config)
 
-    device_list = tuya.get_all_devices()
-    load_devices(device_list)
-
-    def poll_devices_update(event_time):
+        for ha_type, dev_ids in device_type_list.items():
+            config_entries_key = f"{ha_type}.tuya"
+            if config_entries_key not in hass.data[CONFIG_ENTRY_IS_SETUP]:
+                hass.data[DOMAIN]["pending"][ha_type] = dev_ids
+                hass.async_create_task(
+                    hass.config_entries.async_forward_entry_setup(entry, ha_type)
+                )
+                hass.data[CONFIG_ENTRY_IS_SETUP].add(config_entries_key)
+            else:
+                async_dispatcher_send(hass, TUYA_DISCOVERY_NEW.format(ha_type), dev_ids)
+
+    device_list = await hass.async_add_executor_job(tuya.get_all_devices)
+    await async_load_devices(device_list)
+
+    async def async_poll_devices_update(event_time):
         """Check if accesstoken is expired and pull device list from server."""
         _LOGGER.debug("Pull devices from Tuya.")
-        tuya.poll_devices_update()
+        await hass.async_add_executor_job(tuya.poll_devices_update)
         # Add new discover device.
-        device_list = tuya.get_all_devices()
-        load_devices(device_list)
+        device_list = await hass.async_add_executor_job(tuya.get_all_devices)
```

Comment:
Please group the executor jobs into a single call as switching contexts is not cheap.

Reply:
All executor job grouping done